rig-core/src/vector_store/mod.rs

```diff
@@ -50,167 +51,78 @@ pub trait VectorStore: Send + Sync {
-        query: &str,
-        n: usize,
-    ) -> impl std::future::Future<Output = Result<Vec<(f64, T)>, VectorStoreError>> + Send {
-        async move {
-            let documents = self.top_n_from_query(query, n).await?;
-            Ok(documents
-                .into_iter()
-                .map(|(distance, doc)| (distance, serde_json::from_value(doc.document).unwrap()))
-                .collect())
-        }
-    }
+    ) -> impl std::future::Future<Output = Result<Vec<(f64, String, T)>, VectorStoreError>> + Send;
 
     /// Same as `top_n_from_query` but returns the document ids only.
-    fn top_n_ids_from_query(
+    fn top_n_ids<T: for<'a> Deserialize<'a> + std::marker::Send>(
         &self,
         query: &str,
         n: usize,
     ) -> impl std::future::Future<Output = Result<Vec<(f64, String)>, VectorStoreError>> + Send
     {
         async move {
-            let documents = self.top_n_from_query(query, n).await?;
-            Ok(documents
-                .into_iter()
-                .map(|(distance, doc)| (distance, doc.id))
-                .collect())
-        }
-    }
-
-    /// Get the top n documents based on the distance to the given embedding.
-    /// The distance is calculated as the cosine distance between the prompt and
-    /// the document embedding.
-    /// The result is a list of tuples with the distance and the document.
-    fn top_n_from_embedding(
-        &self,
-        prompt_embedding: &Embedding,
-        n: usize,
-    ) -> impl std::future::Future<Output = Result<Vec<(f64, DocumentEmbeddings)>, VectorStoreError>> + Send;
-
-    /// Same as `top_n_from_embedding` but returns the documents without its embeddings.
-    /// The documents are deserialized into the given type.
-    fn top_n_documents_from_embedding<T: for<'a> Deserialize<'a>>(
-        &self,
-        prompt_embedding: &Embedding,
-        n: usize,
-    ) -> impl std::future::Future<Output = Result<Vec<(f64, T)>, VectorStoreError>> + Send {
-        async move {
-            let documents = self.top_n_from_embedding(prompt_embedding, n).await?;
-            Ok(documents
-                .into_iter()
-                .map(|(distance, doc)| (distance, serde_json::from_value(doc.document).unwrap()))
-                .collect())
-        }
-    }
-
-    /// Same as `top_n_from_embedding` but returns the document ids only.
-    fn top_n_ids_from_embedding(
-        &self,
-        prompt_embedding: &Embedding,
-        n: usize,
-    ) -> impl std::future::Future<Output = Result<Vec<(f64, String)>, VectorStoreError>> + Send
-    {
-        async move {
-            let documents = self.top_n_from_embedding(prompt_embedding, n).await?;
-            Ok(documents
+            Ok(self
+                .top_n::<T>(query, n)
+                .await?
                 .into_iter()
-                .map(|(distance, doc)| (distance, doc.id))
+                .map(|(distance, id, _)| (distance, id))
                 .collect())
         }
```

Comment:
Remove the provided implementation of `top_n_ids` since this can vary based on the DB